Support built-in function authentication with OpenFaaS IAM.

welteki opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

Add support for built-in function authentication with OpenFaaS IAM to the go-sdk.

To prevent a breaking change by changing the client constructor function signature a new constructor function `NewClientWithOpts` was added. 

```go
gatewayURL, _ := url.Parse("http://127.0.0.1:8080")
auth := &sdk.BasicAuth{
    Username: username,
    Password: password,
}

client := sdk.NewClientWithOpts(gatewayURL, http.DefaultClient, sdk.WithAuthentication(auth))
```

The client has a new method `InvokeFunction` that can be used to invoke OpenFaaS functions.

When the `auth` argument is `true` the function invocation will be authenticated with an OpenFaaS function access token. The `InvokeFunction` method handles exchanging an ID token obtained from the `FunctionTokenSource` for an OpenFaaS function access token with a single function as audience.

Function access tokens are cached to prevent going through the token exchange each time a function is called.

By default the authentication provider set in `ClientAuth` is used as the function token source if it implements the `TokenSource` interface. Alternatively a function token source can be explicitly set with the `WithFunctionTokenSource` option during client construction.

The `ExchangeIDToken` token function accepts an extra list of exchange options to configure the token exchange. Supported options are `WithAudience` and `WithScope`.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [ ] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
- [ ] My issue has received approval from the maintainers or lead with the `design/approved` label

Support built-in function authentication with OpenFaaS IAM.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

These changes have been used and tested E2E in the OpenFaaS CLI, pro connector sdk and dashboard.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)


## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

welteki committed:
Refactor ClientCredentialsTokenSource to use Token struct

Signed-off-by: Han Verstraete (OpenFaaS Ltd) <han@openfaas.com>

welteki committed:
Add scope to token

Signed-off-by: Han Verstraete (OpenFaaS Ltd) <han@openfaas.com>

welteki committed:
Add exchange options

Signed-off-by: Han Verstraete (OpenFaaS Ltd) <han@openfaas.com>

welteki committed:
Have TokenAuth type implement the TokenSource interface

Signed-off-by: Han Verstraete (OpenFaaS Ltd) <han@openfaas.com>

welteki committed:
Add support for invoking functions using the client

Signed-off-by: Han Verstraete (OpenFaaS Ltd) <han@openfaas.com>

welteki committed:
Improve errors for token exchange functions

Signed-off-by: Han Verstraete (OpenFaaS Ltd) <han@openfaas.com>

welteki committed:
Don't modify the client gateway url struct in client methods

Signed-off-by: Han Verstraete (OpenFaaS Ltd) <han@openfaas.com>

welteki committed:
Cache function invocation tokens

Signed-off-by: Han Verstraete (OpenFaaS Ltd) <han@openfaas.com>

welteki committed:
Improve code comments and update README

Signed-off-by: Han Verstraete (OpenFaaS Ltd) <han@openfaas.com>

welteki committed:
Dump exchange request when FAAS_DEBUG is set

Signed-off-by: Han Verstraete (OpenFaaS Ltd) <han@openfaas.com>

welteki committed:
Make function token cache optional in client

Signed-off-by: Han Verstraete (OpenFaaS Ltd) <han@openfaas.com>

welteki committed:
Support configuring the http client for token exchange requests

Signed-off-by: Han Verstraete (OpenFaaS Ltd) <han@openfaas.com>

alexellis commented on tokencache.go:
```diff
@@ -41,3 +50,33 @@ func (c *MemoryTokenCache) Get(key string) (*Token, bool) {
 }
+
+// StartGC starts garbage collection of expired tokens.
+func (c *MemoryTokenCache) StartGC(ctx context.Context, gcInterval time.Duration) {
```
How would a user access this when they've constructed a Client and have started using it?

alexellis replied:
Likewise, how do you stop this?

welteki replied:
You would have to start it when instantiating the cache. The context can be used to stop it.

The cache can be passed in as an option when the client is constructed.

```go
fnTokenCache := sdk.NewMemoryTokenCache()
go fnTokenCache.StartGC(context.Background(), time.Second*10)

client := sdk.NewClientWithOpts(
    gatewayUrl,
    httpClient,
    sdk.WithAuthentication(auth),
    sdk.WithFunctionTokenCache(fnTokenCache),
)
```

alexellis replied:
Is that example in the examples in the README?

How would you change the example so someone could stop the StartGC?

welteki replied:
Yes this example is also in the README.

Instead of passing the background context in the example it can be changed to use a context that can be cancelled.

welteki committed:
Support garbage collection of expired tokens from cache

Signed-off-by: Han Verstraete (OpenFaaS Ltd) <han@openfaas.com>

welteki committed:
Accept a http.Request as argument to InvokeFunction

Allow the caller to construct and configure requests as required without
limitations imposed by the signature of InvokeFunction.

Signed-off-by: Han Verstraete (OpenFaaS Ltd) <han@openfaas.com>

alexellis reviewed (approved):
Apprved
